lib/logstash/integrations/internal/input.rb

```diff
@@ -0,0 +1,48 @@
+class ::LogStash::Integrations::Internal::Input < ::LogStash::Inputs::Base
+  config_name "internal"
+
+  config :address, :validate => :string, :required => true
+
+  def register
+    # May as well set this up here, writers won't do anything until
+    # @running is set to false
+    @running = java.util.concurrent.atomic.AtomicBoolean.new(false)
+    listen_successful = ::LogStash::Integrations::Internal.listen(@address, self)
+    if !listen_successful
+      raise ::LogStash::ConfigurationError, "Internal input at '#{@address}' already bound! Addresses must be globally unique across pipelines."
+    end
+  end
+
+  def run(queue)
+    @queue = queue
+    @running.set(true)
+
+    while @running.get()
+      sleep 0.5
+    end
+  end
+
+  def running?
+    @running && @running.get()
+  end
+
+  # Returns false if the receive failed due to a stopping input
+  # To understand why this value is useful see Internal.send_to
+  def internal_receive(events)
+    return false if !@running.get()
+
+    events.each do |e| 
+      clone = e.clone
```

Comment:
btw the 5.x logstash event doesn't clone the metadata:

```
1.9 :011 > e = LogStash::Event.new
 => 2017-08-24T21:33:14.106Z %{host} %{message} 
1.9 :012 > e.set("[@metadata][hey]", 1)
 => 1 
1.9 :013 > e.clone.get("[@metadata][hey]")
 => nil 
```

issue about it here: https://github.com/elastic/logstash/issues/4640